com.ca.lsp.cobol/lsp-core-cobol-parser/src/test/java/com/ca/lsp/core/cobol/visitor/MarginABTest.java

```diff
@@ -0,0 +1,191 @@
+          + "000000       CALL 'ATCDEM3'\n"
+          + "000000     END-PERFORM\n"
+          + "000000     IF TAPARM2 = 0\n"
+          + "000000       PERFORM PROCA\n"
+          + "000000     END-IF\n"
+          + "000000  PERFORM WITH TEST\n" // starting area A instead of B
+          + "000000  BEFORE UNTIL TAPARM2 = 0\n" // starting area A instead of B
+          + "000000       SUBTRACT 1 FROM TAPARM2\n"
+          + "000000     END-PERFORM.\n"
+          + "000000  END PROGRAM FILETOTEST.";
+
+  private static final String TEXT_PROGRAM_ID =
+      "000000 IDENTIFICATION DIVISION.\n"
+          + "000000  PROGRAM-ID.    FILETOTEST.\n"
+          + "000000  AUTHOR. SERGIU ILIE.\n"
+          + "000000 DATA DIVISION.\n"
+          + "000000  WORKING-STORAGE SECTION.\n"
+          + "000000  01 TASTRUCT.\n"
+          + "000000     05 LOC-ID.\n"
+          + "000000         10 STATE    PIC X(2).\n"
+          + "000000         10 CITY     PIC X(3).\n"
+          + "000000     05 OP-SYS     PIC X(3).\n"
+          + "000000 PROCEDURE DIVISION.\n"
+          + "000000     MOVE 'ILCHIMVS' TO TASTRUCT.\n"
+          + "000000     MOVE 'ILSPR' TO LOC-ID.\n"
+          + "000000     MOVE 'AIX' TO OP-SYS.\n"
+          + "000000  END PROGRAM FILETOTEST2."; // PROGRAM-ID must be similar with PROGRAM-ID
+  // mentioned on ID DIVISION
+
+  /** In TEXT_DECLARATIVES the mistake is on the DECLARATIVE line */
+  private static final String TEXT_DECLARATIVES =
+      "000000 IDENTIFICATION DIVISION.\n"
+          + "000000  PROGRAM-ID.    FILETOTEST.\n"
+          + "000000  AUTHOR. SERGIU ILIE.\n"
+          + "000000 DATA DIVISION.\n"
+          + "000000  WORKING-STORAGE SECTION.\n"
+          + "000000  01 TASTRUCT.\n"
+          + "000000     05 LOC-ID.\n"
+          + "000000         10 STATE    PIC X(2).\n"
+          + "000000         10 CITY     PIC X(3).\n"
+          + "000000     05 OP-SYS     PIC X(3).\n"
+          + "000000 PROCEDURE DIVISION.\n"
+          + "000000 DECLARATIVES. MAMA\n" // after DECLARATIVES keyword no other token is allowed
+          + "000000 COBOL-DEBUG SECTION.\n"
+          + "000000     USE FOR DEBUGGING ON ALL PROCEDURES.\n"
+          + "000000 COBOL-DEBUG-PARA.\n"
+          + "000000     IF READY-TRACE THEN\n"
+          + "000000       DISPLAY DEBUG-NAME\n"
+          + "000000     END-IF.\n"
+          + "000000  END DECLARATIVES.\n"
+          + "000000     MOVE 'ILCHIMVS' TO TASTRUCT.\n"
+          + "000000     MOVE 'ILSPR' TO LOC-ID.\n"
+          + "000000     MOVE 'AIX' TO OP-SYS.\n"
+          + "000000  END PROGRAM FILETOTEST.";
+
+  @Test
+  public void checkForAreaA() {
+    CobolLanguageEngine engine =
+        Guice.createInjector(new EngineModule(), new DatabusModule())
+            .getInstance(CobolLanguageEngine.class);
+    ResultWithErrors<SemanticContext> result;
+
+    result = engine.run("1", TEXT_DIVISION_WRONG_PLACE, "DID_OPEN");
+    assertEquals(3, result.getErrors().size());
```

Comment:
You may use `UseCaseEngine#analyze` or `UseCaseEngine#analyzeForErrors` here to reduce code duplication